Feature/sync video

pajakgdynia opened the pull request:
A "bit" refreshed videoReader, videoPlayer, thumbnailer.
All frames are synchronized with courses (distance to frame mapping).
"Strange" formats works fine (starting frame not 0, frames numbered not by 1, frames with duration).
Several simplification.
I thought it will be much simpler.. but life is cruel :)

pajakgdynia commented:
aaargh. As far as I saw.. There are some changes which shouldn't be merged. You have to read the code :+1:

ibooij commented:
I've tried merging this, see https://github.com/ibooij/big-ring/tree/pajakgdynia-feature/sync_video, but I found several errors that I don't know how to fix:

* If I select a course which does not start at the beginning of the video file, the videos starts "fast-forwarding" during cycling. For some larger videos this takes quite a while. Shouldn't it do a proper seek in this case before signalling that the video is ready?
* When running in debug mode, some asserts are triggered when loading a video, for instance Q_ASSERT(_timeDiff > 0); in genericvideoreader.cpp on line 312. _timeDiff is -1 when loaded.
Also   Q_ASSERT(sizeof(AVFrame::linesize) == sizeof(uint8_t *) * AV_NUM_DATA_POINTERS); on in genericvideoreader.cpp on line 285.

Would you be able to fix these? 

cheers,
Ilja

pajakgdynia commented:
> If I select a course which does not start at the beginning of the video
file, the videos starts "fast-forwarding" during cycling. For some larger
videos this takes quite a while.
I don't understand.. Does the video freeze? Or is it played (much) faster
than expected?

Show me these "funny lines" from the log file (with numbers and dashes). In
first line you can see how many frames were skip during each step, second
shows the time to get the frame from the cache (0 and '-' are ok, other
values suggest you have to slow computer to play this video..), third line
is time diff between steps (periodicity).

> Shouldn't it do a proper seek in this case before signalling that the
video is ready?

It does (almost the same way as before). It MUST start exactly from the
frame shown as a thumbnail. First frame is shown for about 2s (Simulation
initialization time?). During this time frames are only cached, not
displayed. When time starts, video should go forward (with expected speed).

Could you give me a name of the video you tried to run (and/or where is it
taken from) to see what happen?

> When running in debug mode, some asserts are triggered when loading a
video, for instance Q_ASSERT(_timeDiff > 0); in genericvideoreader.cpp on
line 312. _timeDiff is -1 when loaded.

Hmm.. Doesn't sound good. timeDiff is computed on first (video) packets,
some have to be buffered to complete the frame. libAV reads more packets to
complete first frame (4, 5 or 6 packets are read). timeDiff is set on
second packet (in general timeDiff is 1, but I saw 10, 100, 400).
Enable loggers from genericVideoReader (comment out lines 16/17) and send
me logs, please.


> Also Q_ASSERT(sizeof(AVFrame::linesize) == sizeof(uint8_t *) *
AV_NUM_DATA_POINTERS); on in genericvideoreader.cpp on line 285.

My fault.. It is sanity checking only (for memcopies).. Somehow I merged
two lines (by mistake?), and didn't noticed (pointer size and int have same
size on 32bit..). They should be:

>     Q_ASSERT(sizeof(AVFrame::linesize) == AV_NUM_DATA_POINTERS *
sizeof(int));
>     Q_ASSERT(sizeof(AVFrame::data) == AV_NUM_DATA_POINTERS *
sizeof(uint8_t *));

Or even better is to shift them before line 272.

pajakgdynia commented:
Ok, "fast-forwarding".. There's no 2s sleep (why? it was ok), so frames are cached just after start (you can see small freeze and then speed-up). I've tried it with 500W constant (it gains speed faster). Behaves as expected (ok, as I expect :) )

I haven't seen any issues related timeDiff.. Logs from GenericVideoReader still expected (from linux, I assume).

ibooij commented:
Hi,

I tried recreating the error. What is especially interesting in the log below is the where I opened JP_Aso_San. I tried to go the a (self defined) segment at about 33km. When I started that the video was placed in the beginning and it started going through the video at very high pace. I stopped that ride and went back. Selected the same segment again and started again. Now it started at the right place.

I'm beginning to think that the video reader is still busy doing other things and gets into a wrong state. But that's just a thought. 

It would probably also help me if I understood your video reader a bit better. That will take some time, but perhaps you could also document it a bit more. As it is multi-threaded, which is always a bit tricky in my opinion (but sometimes necessary!), thorough documentation would certainly help.

Again, thanks for your time, here is the log: https://gist.github.com/ibooij/3cff6b2e46f441ec42dd

ibooij commented on mainlib/videoinforeader.cpp:
```diff
@@ -14,10 +14,12 @@ VideoInfoReader::~VideoInfoReader()
+    //AVStream* stream = videoStream();
+    //VideoInfo videoInfo = { static_cast<float>(av_q2d(stream->avg_frame_rate)) };
+    //return videoInfo;
+    VideoInfo videoInfo = { 25.0 };
```
Does GPX import still work now? It needs to know the framerate of the video to know what framenumbers go with what distance.

ibooij commented:
It must be said, the video does run very smooth, even when the framerate is high!

pajakgdynia commented:
In your log.. I've taken a look again, and now I see the problem:

> [11:41:34] DEBUG [void Thumbnailer::videoOpened(const VideoInfo*):116 @0x7f02614ec880] File closed while requested "/mnt/windows/Users/Ilja/Documents/RLV/videos/ES_Teide_2012.avi"
> ...
> [11:41:34] DEBUG [void VideoPlayer::setVideoOpened(const VideoInfo*):63 @0x7f02614ec880] Cannot open video
> ...
> [11:41:34] DEBUG [void VideoPlayer::seekToFrame(quint32):90 @0x7f02614ec880] Seek to frame 145230
> ...
> [11:41:34] DEBUG [void VideoPlayer::setSeekReady(qint64):99 @0x7f02614ec880] Sought to frame 0

Somehow video was not opened immediatelly (?? why it took so long??) and it gave strange results. It is wrong behaviour ("cannot open video" should not happen here)! For sure opening a video file by detailsPanel would solve the problem.. but again, it must work even now (each panel tries to open the file itself).

pajakgdynia commented:
4 lines of log.. and so many corrections were necessary. It should work much better now (works with sleep(10) during open..). Merge it, please.

ibooij commented:
I'll take a longer look the coming weekend. I really want to know what we're merging. I'll keep you updated.

ibooij commented:
I've checked out the HEAD of your branch and merged "my" master to it. 
I fixed one major regression: There was a `videoCorrection` value in class RealLifeVideo, which was needed for older Tacx RLVs (like FR_AlpineClassic_07 and NL_AmstelGoldRace) which have video files that are shorter or longer than the distance mappings in the rlv files suggest. I've put this videoCorrection back in.

The current result is in branch pajakgdynia-feature/sync_video2. 

It mostly works but I'm seeing some crashes when requesting different thumbnails. Something with _diffTime in the GenericVideoReader being smaller then 0 and triggering an ASSERT.

I'm also seeing some old C-style casts `(T)` instead of C++ casts `static_cast<T>`, which I do not really like. 

The main thing is, though, that I don't really understand what is happening in the video classes, especially with regards to the threads and the messages passed back and forth. Would you be able to write or draw an explanation? I don't want to try and fix crashes or style issues without understanding it. Perhaps we could make the code somewhat simpler also. That would also make it easier to spot the location of the crashes I'm seeing.

thanks,
Ilja

ibooij commented:
Found another thing. VideoInfoReader does not work anymore. 

Things still work because the framerate is read later on, I guess (not sure), but this is not really the most clear way I would think.

pajakgdynia commented:
:)

> I fixed one major regression: There was a videoCorrection value in class
> RealLifeVideo, which was needed for older Tacx RLVs (like
> FR_AlpineClassic_07 and NL_AmstelGoldRace) which have video files that are
> shorter or longer than the distance mappings in the rlv files suggest. I've
> put this videoCorrection back in.
>
> Are you sure it is proper solution? Are there any comments from TacX or
RLV forums?

The problem exists only for videos with frame-distance mapping. It doesn't
for time-distance (.gpx, "old" cycleOps virtualTrainng).
I do understand, that it is the issue related only to .rlv videos. Why
don't correct it only there? For sure, it should not be a "common" code.

I saw that some video are badly synchronized (I mean.. RLV file has 10%
uphill.. while in video you can see downhill..) In some cases there is an
offset, in some framerate is incorrectly specified. When video has been
filmed with constant speed (regardless the slope), it matches second case.

But still there's no "general" solution. There should be available
"Advanced properties" for route, with offset, FPS correction (calculated
automatically, but still available for the user) and route length (to
trim/extend the route/courses)

I put closer look on these two videos.

_diffTime in the GenericVideoReader being smaller then 0 and triggering an
> ASSERT.
>
timeDiff condition is very important (it shows how consecutive frames are
numbered). "Our" frame number is decoded using this value. It usually is 1,
but I saw several other values (10, 100, 400, I wonder what for :) ).
It can happen only if somehow (how?) you request the frame of closed video
(then is -1, "uninitialized").
Does it still happen? Will you tried it with "my current" version?
In "my current" branch I corrected videoReader a bit, it is more
"resistant" to user actions ("fast clicking": changing videos, courses,
etc). I noticed one core condition there..

I'm also seeing some old C-style casts (T) instead of C++ casts
> static_cast<T>, which I do not really like.
>
My fault. Keep in mind I'm C programmer :)
There are four types:
- signed/unsigned casts, some values are signed for being uninitialized;
mostly in conditions (it would look ugly)
- increasing resolution, some computations must be done on doubles.
Additional variables shall be used..
- sometimes frames should be 64bits.. There is a mess with these, it should
be discussed.. and cleaned.
- "plasma" enums (events). Shall I do it for you? :)

> The main thing is, though, that I don't really understand what is
> happening in the video classes, especially with regards to the threads and
> the messages passed back and forth. Will you be able to write or draw an
> explanation? I don't want to try and fix crashes or style issues without
> understanding it.
>
The idea IS very simple.
Additional thread ("worker") deals with libAV, main thread ("reader") keeps
frames. Events are passed to/from "worker", slots/signals are handled by
"reader".
GenericReader.. more or less is a bridge between "our" code and libAV. In
(almost) same way as before.
"Worker" encapsulates genericReader plain requests, it deals with all these
DTSes, wrong frames, frames with duration.
"Reader" interface is pretty simple, event driven style: just several
commands (Open, Request Frame / Image, Close), they emit signals (Opened,
Frame / Image Available).
The most sophisticated thing in the "reader" is cache and reading frames in
"read ahead" mode. Conditions when (and where) to seek are quite complex.
And the last, and again complex.. It tries to rid of race conditions.
Reading videos is terribly slow (seeking might take even several seconds),
and.. issues must be hidden there.

I'm going to solve one more issue: seeking FORWARD and reporting frame with
big duration.. It solves resyncing problem with slow computers (decoding
big GOPs..)

> Perhaps we could make the code somewhat simpler also. That would also make
> it easier to spot the location of the crashes I'm seeing.
>
> I've though.. that it will be simpler.. But I was wrong :) Division makes
spotting easier. GenericReader => libAV, worker => video (or libAV), reader
=> my fault :( You can better diagnose the issue with loggers enabled
(comment out qDebug() macros)


> Found another thing. VideoInfoReader does not work anymore.
>
> Things still work because the framerate is read later on, I guess (not
> sure), but this is not really the most clear way I would think.
>
Even more.. Does not exist anymore. VideoInfo is emited with VideoOpened
signal. There is a bit more than in "original" record.
VideoInfo was used only for .gpx files.. to compute frame (computed from
position time) to distance mapping. You should notice that there are two
framerates: declared in control file (to compute proper mapping) and video.
They mustn't equal! I'll create new issue for related bug.

ibooij commented:
> :)
> >  I fixed one major regression: There was a videoCorrection value in class
> >  RealLifeVideo, which was needed for older Tacx RLVs (like
> >  FR_AlpineClassic_07 and NL_AmstelGoldRace) which have video files that are
>  >shorter or longer than the distance mappings in the rlv files suggest. I've
>  > put this videoCorrection back in.

> Are you sure it is proper solution? Are there any comments from TacX or
> RLV forums?

I got this information from Klaudiusz from  [RLV-Workbench](http://rlv-stormarn.de/). It does work correctly. For instance, for FR_AlpineClassic_07, I knew the start location of the Alpe d'Huez TT, but not using this correction it was completely in the wrong place (about 1km before the real start). After correction it is at the right place. 

> 
> The problem exists only for videos with frame-distance mapping. It doesn't
> for time-distance (.gpx, "old" cycleOps virtualTrainng).
> I do understand, that it is the issue related only to .rlv videos. Why
> don't correct it only there? For sure, it should not be a "common" code.

That is correct. However, then we would have to open the video file for every rlv file that is imported, which would take quite a while. There is of course the plan to really import the rlvs and load from an internal storage format every time after that, but until we have that, this is a working solution that should not be broken.

> 
> I saw that some video are badly synchronized (I mean.. RLV file has 10%
> uphill.. while in video you can see downhill..) In some cases there is an
> offset, in some framerate is incorrectly specified. When video has been
> filmed with constant speed (regardless the slope), it matches second case.
> 
> But still there's no "general" solution. There should be available
> "Advanced properties" for route, with offset, FPS correction (calculated
> automatically, but still available for the user) and route length (to
> trim/extend the route/courses)

I do not want to bother any users with "Advanced" options. Users should just be able to use the application and we will have to make sure that they can.
> 
> I put closer look on these two videos.
> 
> EDIT. Ok I do understand the issue. The problem is between mappings, thus "correction" code must be in RlvParser. And if length for profile and video differs (too much.. let say 3s at average speed), then route shall not be importable.
> 
> Egs.. Lengths from profile/mapping (from free cycling rlv):
> "ES_cruzverde" 2.51479e+09/14929.5
> "Wiesbaden90k" 91037.8/89894
> The question is.. which is proper, and which must be corrected and how..

I think we should do our best to make the rlvs work. I'm happy that my old Tacx rlvs still work with the software.

>>  _diffTime in the GenericVideoReader being smaller then 0 and triggering an
>  ASSERT.

> timeDiff condition is very important (it shows how consecutive frames are
> numbered). "Our" frame number is decoded using this value. It usually is 1,
> but I saw several other values (10, 100, 400, I wonder what for :) ).
> It can happen only if somehow (how?) you request the frame of closed video
> (then is -1, "uninitialized").
> Does it still happen? Will you tried it with "my current" version?
> In "my current" branch I corrected videoReader a bit, it is more
> "resistant" to user actions ("fast clicking": changing videos, courses,
> etc). I noticed one core condition there..
> 
> EDIT. I've found troubleshooting video. If single packet makes the frame.. Assert will be triggered.
> I've corrected it, but some videos (with different timeDiff) won't work either.. I need to find better solution..

If you have a good solution, could you add it to your sync_video branch? 

>>  I'm also seeing some old C-style casts (T) instead of C++ casts
>>  static_cast<T>, which I do not really like.

> My fault. Keep in mind I'm C programmer :)

I'm a Java/C/C++/Python programmer myself, so I always have to stop myself from putting constructs from one language into the other.. I'll change the casts when I see some in some code I'm toying with.


> There are four types:
> - signed/unsigned casts, some values are signed for being uninitialized;
> mostly in conditions (it would look ugly)
> - increasing resolution, some computations must be done on doubles.
> Additional variables shall be used..
> - sometimes frames should be 64bits.. There is a mess with these, it should
> be discussed.. and cleaned.
Yep, I guess we should just be using 64 bit integers everywhere for ints and perhaps doubles for floating point values.
> - "plasma" enums (events). Shall I do it for you? :)

>>  The main thing is, though, that I don't really understand what is
>>  happening in the video classes, especially with regards to the threads and
>>  the messages passed back and forth. Will you be able to write or draw an
>>  explanation? I don't want to try and fix crashes or style issues without
>>  understanding it.

> The idea IS very simple.
> Additional thread ("worker") deals with libAV, main thread ("reader") keeps
> frames. Events are passed to/from "worker", slots/signals are handled by
> "reader".
> GenericReader.. more or less is a bridge between "our" code and libAV. In
> (almost) same way as before.
> "Worker" encapsulates genericReader plain requests, it deals with all these
> DTSes, wrong frames, frames with duration.
> "Reader" interface is pretty simple, event driven style: just several
> commands (Open, Request Frame / Image, Close), they emit signals (Opened,
> Frame / Image Available).
> The most sophisticated thing in the "reader" is cache and reading frames in
> "read ahead" mode. Conditions when (and where) to seek are quite complex.
> And the last, and again complex.. It tries to rid of race conditions.
> Reading videos is terribly slow (seeking might take even several seconds),
> and.. issues must be hidden there.

Ok. I was trying some stuff yesterday evening and things became more clear. I changed the cache to use std::unique_ptr instead of raw pointers, to make ownership more clear. Also, I moved some classes to their own files, to be able to distinguish the code a bit better. I like small files. Does that make me a typical Java programmer..? ;-)

What sometimes seems to go wrong, especially when reading a video is slow, is that an `assert` in the VideoReaderWorker is triggered. Perhaps because the wrong video file is opened at that moment.

Because the worker is running on a different thread, we have to assume it might get requests that it does not expect at that moment. The same goes for the VideoReaderThreaded, which should be able to cope with responses coming in after a completely different request has come in. 

There is already some code to handle this, but there are also asserts triggered when things go wrong. I'm wondering if I'm seeing more of those because I'm using two computers with 2 cores and hyperthreading. That would make it easier to have threading issues.

> 
> I'm going to solve one more issue: seeking FORWARD and reporting frame with
> big duration.. It solves resyncing problem with slow computers (decoding
> big GOPs..)

Good, please put it on your feature/sync_video branch.

>>  Perhaps we could make the code somewhat simpler also. That would also make
>>  it easier to spot the location of the crashes I'm seeing.

> I've though.. that it will be simpler.. But I was wrong :) Division makes
> spotting easier. GenericReader => libAV, worker => video (or libAV), reader
> => my fault :( You can better diagnose the issue with loggers enabled
> (comment out qDebug() macros)
> >  Found another thing. VideoInfoReader does not work anymore.
> 
>  Things still work because the framerate is read later on, I guess (not
>  sure), but this is not really the most clear way I would think.
> Even more.. Does not exist anymore. VideoInfo is emited with VideoOpened
> signal. There is a bit more than in "original" record.
> VideoInfo was used only for .gpx files.. to compute frame (computed from
> position time) to distance mapping. You should notice that there are two
> framerates: declared in control file (to compute proper mapping) and video.
> They mustn't equal! I'll create new issue for related bug.

Then VideoInfoReader should be completely removed. It is a "fake" class now.

I don't understand why the framerate in the video file and in the control file should not be equal. Could you explain?

cheers,
Ilja 
I

pajakgdynia commented:
>> If single packet makes the frame.. Assert will be triggered.
>> I've corrected it, but some videos (with different timeDiff) won't work either.. I need to find better solution..
> If you have a good solution, could you add it to your sync_video branch?
I'll try to move commits.. I run "bigger picture", these changes are "side changes" only. 

> I changed the cache to use std::unique_ptr instead of raw pointers
I still cannot think in C++/C++11 subjects.. Good move :)
Now I started thinking about changing raw poiters (RealLifeVideo*) to something else. A lot of them is in my code..

> Then VideoInfoReader should be completely removed. It is a "fake" class now.
It is removed, the only question might be where ("import_videos" branch). I realized it too late.

> I don't understand why the framerate in the video file and in the control file should not be equal. Could you explain?
Rather "could". Most RLVs have 25 FPS. Some videos don't (they are 50 FPS). Video could have been recoded, or RLV file was made in the software with "constant FPS"/"assumed FPS".

ibooij commented:
>> I don't understand why the framerate in the video file and in the control file should not be equal. Could you explain?

> Rather "could". Most RLVs have 25 FPS. Some videos don't (they are 50 FPS). Video could have been recoded, or RLV file was made in the software with "constant FPS"/"assumed FPS".

Ok, then I understand. Weird that some videos have a different framerate in the control file then in the video, but very good that you managed to cope with it in this way.

ibooij commented:
I have found a cause for the `Q_ASSERT(_timeDiff > 0);` crash. 

I have (at least one) video that has the complete first frame in the first packet. In `GenericVideoReader::loadNextFrame()` the value for `_timeDiff` is calculated from the difference between two packets. If the frame comes from the first packet, no value is calculated. 

I added to the last part of `GenericVideoReader::loadNextFrame()` to use a default value for _timeDiff if there is no other way to get the value.
```
// if we had a frame in the first packet, we'll initialize
// _timeDiff to 1 because we don't know any better.
if (_timeDiff == -1) {
    _timeDiff = 1;
}
```

pajakgdynia commented:
Yes, exactly same thing I've done (ok, with warning message, to see if something is wrong in spite of disabled debugs).
And.. Exactly only one video with such issue. SvelvikveienTT?

ibooij commented:
For me it was BE_NamurRevin.

I have some videos which don't work yet with the new player, for instance IT_Siniscola, which you can download from https://www.reallifevideo.de/rlv.php?rlv=albo. It fails to find a key frame I guess.

ibooij commented:
Not going to merge this as is, so closing.
